### PR TITLE
Practice tutorials

### PR DESCRIPTION
**Pull request by Avinashvytla351**

### CODE REVIEW

File: Client/index.js [hunk 1]
```diff
@@ -2322,6 +2456,121 @@ app.get("/codechef-iare-chapter", async (req, res) => {
...
+      });
+    } else {
+      res.render("error", {
+        data: body,
+        imgUsername: req.cookies.username,
+      });
+    }
+  });
+});
+
```
Comment (nightsailor) [inline Client/index.js:2573]:
```suggestion
```

File: Client/index.js [hunk 2]
```diff
@@ -37,6 +37,8 @@ app.use("/ide", express.static(path.resolve("../IDE")));
 
 app.use("/tutorials", express.static(__dirname + "/"));
 app.use("/contests", express.static(__dirname + "/"));
+app.use("/courses", express.static(__dirname + "/"));
+app.use("/courses/contest", express.static(__dirname + "/"));
```
Comment (nightsailor) [inline Client/index.js:41]:
```suggestion
```

**Comment by Avinashvytla351** [comment]
Please Review

File: Server/routes/question.route.js [hunk 3]
```diff
@@ -26,21 +26,35 @@ module.exports = (app) => {
...
     questions.createTutorials
   );
 
+  // Create a new question for Practice
+  app.post(
+    "/questionpractice",
+    middleware.checkTokenAdmin,
+    questions.createTutorials
+  );
+
```
Comment (nightsailor) [inline Server/routes/question.route.js:43]:
```suggestion
```

**Comment by Avinashvytla351** [comment]
please check the changes

File: Client/index.js [hunk 4]
```diff
@@ -2052,6 +2085,69 @@ app.get("/tutorials/:courseId/progress", async (req, res) => {
...
+        body.courseName = "Invalid Course";
+      }
+      res.render("displayTutQuestions", {
+        imgUsername: req.cookies.username,
+        data: body,
+      });
+    });
+  });
+});
+
```
Comment (nightsailor) [inline Client/index.js:2150]:
```suggestion
```

File: Client/index.js [hunk 5]
```diff
@@ -2136,19 +2231,39 @@ app.get("/tutorials/:courseId/:difficulty/:concept", async (req, res) => {
...
+        "/" +
+        req.params.difficulty,
+      method: "get",
+      headers: {
+        authorization: req.cookies.token,
+      },
+      json: true,
+    };
+  }
   request(options, function (err, response, body) {
```
Comment (nightsailor) [inline Client/index.js:2267]:
```suggestion
  let difficulty = req.params.difficulty;

  let categoryBased = difficulty.includes("-");
  let ifTopicsOrCompanies =
    req.params.difficulty === "Topics" || req.params.difficulty === "Companies";

  let param = ifTopicsOrCompanies
    ? "topics"
    : categoryBased
    ? difficulty.split("-").join("/")
    : difficulty;

  let options = {
    url:
      serverRoute +
      "/questions/" +
      (categoryBased ? "practice/" : "courses/") +
      req.params.courseId +
      "/" +
      param,
    method: "get",
    headers: {
      authorization: req.cookies.token,
    },
    json: true,
  };
  request(options, function (err, response, body) {
    if (ifTopicsOrCompanies) {
      let courseIds = ["IARE_PY", "IARE_C", "IARE_JAVA", "IARE_CPP"];
      let isCourseValid = courseIds.includes(req.params.courseId);
      let ifTopics = difficulty === "Topics";

      body.courseId = req.params.courseId;
      body.courseName = isCourseValid
        ? ifTopics
          ? "Select a topic"
          : "Select a company"
        : "Invalid Course";

      res.render("practiceTutList", {
        imgUsername: req.cookies.username,
        title: difficulty,
        data: body,
      });
    } else {

```

File: Client/index.js [hunk 6]
```diff
@@ -2189,10 +2304,28 @@ app.get("/tutorials/:courseId/:difficulty", async (req, res) => {
...
+          practiceTitle = " Select a company";
+        }
+        body.courseName = practiceTitle;
+        res.render("practiceTutList", {
+          imgUsername: req.cookies.username,
+          title: req.params.difficulty,
+          data: body,
+        });
+      }
     });
```
Comment (nightsailor) [inline Client/index.js:2329]:
```suggestion
  }
```

Comment (nightsailor) [inline Client/index.js:2329]:
```suggestion
      res.render("displayTutQuestions", {
        imgUsername: req.cookies.username,
        data: body,
      });
    });
  }
```

File: Client/views/practiceTutList.ejs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+      <%  myList = myList.concat(myList,data[i].topic);%>
+      <% } else { %>
+      <% myList = myList.concat(myList,data[i].company);%>
+      <%} %>
+      <%} %>
+      <%const unique =new Set(myList); %> 
+      <% var myArr = Array.from(unique) %>
+      <% for(var i=0; i<myArr.length;i++) {%>
+      <a
+        href="<%= data.clientRoute %>/tutorials/practice/<%= data.courseId %>/<%= title%>/<%= myArr[i] %>"
```
Comment (nightsailor) [inline Client/views/practiceTutList.ejs:104]:
```suggestion
        href="<%= data.clientRoute %>/tutorials/practice/<%= data.courseId %>/<%= title%>-<%= myArr[i] %>"
```

File: Client/views/practiceTutList.ejs [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+      <% var myList = [] %>
+      <% for(var i =0;i<data.length;i++) { %>
+      <%   if(title == "Topics") { %>
+      <%  myList = myList.concat(myList,data[i].topic);%>
+      <% } else { %>
+      <% myList = myList.concat(myList,data[i].company);%>
+      <%} %>
+      <%} %>
+      <%const unique =new Set(myList); %> 
+      <% var myArr = Array.from(unique) %>
```
Comment (nightsailor) [inline Client/views/practiceTutList.ejs:101]:
```suggestion
      <% var myList = [] %>
      <% if(title == "Topics") { %>
        <% for(var i =0;i<data.length;i++) { %>
          <%  myList = myList.concat(myList,data[i].topic);%>
      <% } } else { %>
        <% for(var i =0;i<data.length;i++) { %>
          <% myList = myList.concat(myList,data[i].company);%>
      <%} }%>
      <%const myArr = myList.filter((v, i, a) => a.indexOf(v) === i) %>
```

File: Server/controllers/question.controller.js [hunk 9]
```diff
@@ -777,8 +885,64 @@ exports.findAllCourseDifficulty = (req, res) => {
...
+            message: "Question not found with id " + req.params.questionId,
+          });
+        }
+        return res.status(500).send({
+          success: false,
+          message: "Error retrieving question with id " + req.params.questionId,
+        });
+      });
+  }
+};
```
Comment (nightsailor) [inline Server/controllers/question.controller.js:943]:
```suggestion
exports.findAllCourseTopicWise = (req, res) => {
  let param = req.params.title? "topic": "company";
  Question.find({
    courseId: req.params.courseId,
    [param]: req.params.name,
  })
    .then((question) => {
      if (!question) {
        return res.status(404).send({
          success: false,
          message: "Question not found with id " + req.params.questionId,
        });
      }
      res.send(question);
    })
    .catch((err) => {
      if (err.kind === "ObjectId") {
        return res.status(404).send({
          success: false,
          message: "Question not found with id " + req.params.questionId,
        });
      }
      return res.status(500).send({
        success: false,
        message: "Error retrieving question with id " + req.params.questionId,
      });
    });
};
```

**Pull request merged**